### PR TITLE
Fix replication tree with sharding

### LINKED ISSUE #663: Fix replication tree when sharding is applied
### Describe the problem you discovered

P2P message encoding about replication tree doesn't support currently different number of nodes for chain & beacon replication tree, causing the messages to crash.

As replicas size can be different based on the type of transaction (for example, network transactions chain replicas are all the nodes)

### Describe the solution you'd like

We should encode separately the replication tree of beacon & chain replicas.

### PR DESCRIPTION
**Pull request by samuelmanzanera**
# Description

This PR aims to resolve a problem of serialization when the size of the replication tree is different from the chain replicas & beacon replicas

Fixes #663 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

You can apply this patch, with more than 5 running nodes

```diff
diff --git a/lib/archethic/election/constraints/storage.ex b/lib/archethic/election/constraints/storage.ex
index 02ed0edc..7f8fb879 100755
--- a/lib/archethic/election/constraints/storage.ex
+++ b/lib/archethic/election/constraints/storage.ex
@@ -13,8 +13,8 @@ defmodule Archethic.Election.StorageConstraints do
 
   alias Archethic.P2P.Node
 
-  @default_min_geo_patch 4
-  @default_min_geo_patch_avg_availability 0.8
+  @default_min_geo_patch 1
+  @default_min_geo_patch_avg_availability 0.0
 
   @type min_geo_patch_fun() :: (() -> non_neg_integer())
   @type min_geo_patch_avg_availability_fun() :: (() -> float())
@@ -108,9 +108,14 @@ defmodule Archethic.Election.StorageConstraints do
   """
   @spec hypergeometric_distribution(list(Node.t())) :: pos_integer()
   def hypergeometric_distribution(nodes)
-      when is_list(nodes) and length(nodes) >= 0 and length(nodes) <= 10,
-      do: nodes
+      when is_list(nodes) and length(nodes) >= 0 and length(nodes) <= 10 do
+    if length(nodes) > 5 do
+      length(nodes) - 2
+    else
+      length(nodes)
+    end
+  end
 
-  def hypergeometric_distribution(nodes) when is_list(nodes) and length(nodes) >= 0,
+  def hypergeometric_distribution(nodes) when is_list(nodes),
     do: HypergeometricDistribution.run_simulation(length(nodes))
 end

```

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by apoorv-2204** [comment]
How to ?

**Comment by apoorv-2204** [comment]
`+  @default_min_geo_patch_avg_availability 0.0 `?

**Comment by apoorv-2204** [comment]
`@default_min_geo_patch_avg_availability 0.0`?

**Comment by samuelmanzanera** [comment]
> `@default_min_geo_patch_avg_availability 0.0`?

This is to used to say we don't expect minimal availability ratio per geo patch.

File: test/archethic/p2p/messages_test.exs [hunk 1]
```diff
@@ -219,6 +219,55 @@ defmodule Archethic.P2P.MessageTest do
...
+            unspent_outputs: []
+          },
+          signature:
+            <<231, 4, 252, 234, 6, 126, 91, 87, 41, 70, 76, 220, 116, 238, 128, 189, 94, 124, 207,
+              90, 32, 143, 239, 153, 101, 148, 189, 125, 25, 235, 20, 207, 168, 10, 86, 59, 14,
+              249, 104, 144, 141, 151, 232, 149, 24, 189, 225, 56, 65, 208, 220, 202, 169, 166,
+              36, 248, 98, 108, 241, 114, 47, 102, 176, 212>>,
+          protocol_version: ArchethicCase.current_protocol_version()
+        },
+        replication_tree: %{
```
Comment (apoorv-2204) [inline test/archethic/p2p/messages_test.exs:252]: Can we include dynamic creation of replication tree?

Comment (samuelmanzanera) [reply test/archethic/p2p/messages_test.exs:252]: What do you mean ?

**Review by Neylix — APPROVED** [review]

**Review by Neylix — CHANGES_REQUESTED** [review]
In storage constraint if the number of nodes is between 0 and 10 we are returning the list of nodes instead of the number of nodes 
https://github.com/archethic-foundation/archethic-node/blob/438d66e8e6fb21db8195727bf1a657e577d4aee7/lib/archethic/election/constraints/storage.ex#L112

**Comment by samuelmanzanera** [comment]
> In storage constraint if the number of nodes is between 0 and 10 we are returning the list of nodes instead of the number of nodes 
> https://github.com/archethic-foundation/archethic-node/blob/438d66e8e6fb21db8195727bf1a657e577d4aee7/lib/archethic/election/constraints/storage.ex#L112
> 

Yes indeed , to do like in the patch

**Review by Neylix — APPROVED** [review]

**Pull request merged**